lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -810,17 +797,35 @@ getEndPointsOfDataDependentChain(SILValue value, SILFunction *fun,
       llvm::copy(user->getResults(), std::back_inserter(worklist));
     }
   }
-
   // At this point, visitedUsers have all the transitive, data-dependent uses.
-  // Compute the lifetime frontier of all the uses which are the instructions
-  // following the last uses. Every exit from the last uses will have a
-  // lifetime frontier.
+  users.append(visitedUsers.begin(), visitedUsers.end());
+}
+
+/// Collect the end points of the instructions that are data dependent on \c
+/// value. A instruction is data dependent on \c value if its result may
+/// transitively depends on \c value. Note that data dependencies through
+/// addresses are not tracked by this function.
+///
+/// \param value SILValue that is not an address.
+/// \param fun SILFunction that defines \c value.
+/// \param endUsers buffer for storing the found end points of the data
+/// dependence chain.
+static void
+getEndPointsOfDataDependentChain(SILValue value, SILFunction *fun,
+                                 SmallVectorImpl<SILInstruction *> &endUsers) {
+  assert(!value->getType().isAddress());
+
+  SmallVector<SILInstruction *, 16> transitiveUsers;
+  getTransitiveUsers(value, transitiveUsers);
+
+  // Compute the lifetime frontier of all the transitive uses which are the
+  // instructions following the last uses. Every exit from the last uses will
+  // have a lifetime frontier.
   SILInstruction *valueDefinition = value->getDefiningInstruction();
   SILInstruction *def =
       valueDefinition ? valueDefinition : &(value->getParentBlock()->front());
   ValueLifetimeAnalysis lifetimeAnalysis =
-      ValueLifetimeAnalysis(def, SmallVector<SILInstruction *, 16>(
-                                     visitedUsers.begin(), visitedUsers.end()));
+      ValueLifetimeAnalysis(def, transitiveUsers);
```

Comment:
@atrick Just a clarification. Here the ValueLifetimeAnalysis is not used to get the lifetime of OSLogMessage instance nor any other value. It is instead used to find the end points of the (transitive) uses of OSLogMessage so that constant evaluation need not continue beyond those points. The name "lifetimeAnalysis" is a bit misleading. This check is not about lifetimes at all. Bail out early is totally fine, only that we would be missing out on some constant folding opportunities. Transitively following branch dependencies is also fine. We just want an over approximate set of post-domainating points of OSLogMessage beyond which it is not worth constant evaluating. In fact, the "return instructions" of the function (containing OSLogMessage) is a possible over-approximate end point, but I just want to avoid constant evaluator going all the way to the end of the function.

Reply:
Yes, I was trying to contrast your use of VLA with a "normal" use. I would avoid talking about post-dominance here. I would describe the Frontier as known last-use points. i.e. there can't be any data dependencies past that point. I think it's reasonable to bail out on block terminators; obviously you just can't ignore them.